### ISSUE TITLE
How should Chapel handle integer overflow?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#6160]
For example, if you have
``` chapel
var x:int(8) = 127;
x += 1;
writeln(x);
```

what should happen? Let's introduce some terminology for the basic options:

 * *undefined behavior*: this program is wrong, who knows what it will do
  * *wrap*: x is -128
  * *trap*: process dies with a signal (as with, say, dereferencing NULL)
  * *throw*: an overflow is thrown and can be caught 

Note that the main difference in functionality is that *trap* is expected to be unrecoverable but *throw* is explicitly something the user's code can handle.

### What do other languages do?

* C: signed integer overflow is *undefined behavior*; unsigned integer overflow will *wrap*
* [Swift:](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/AdvancedOperators.html) signed and unsigned overflow *trap*
* [Rust](http://huonw.github.io/blog/2016/04/myths-and-legends-about-integer-overflow-in-rust/):
   * in debug mode, "panic" (AFAIK panic maps to *trap* in our terminology since it's not meant to be recovered from; note though a Rust panic does involve stack unwind+cleanup).
   * in release mode, *wrap*
   * further controllable by compiler flags and explicit add functions

### Areas of concern

* performance of generated code
* safety

### Loops and Bounded Ranges

When optimizing loops, it's important to know that the loop induction variables cannot wrap. That is important for optimizations such as loop unrolling or vectorization.

For loops that iterate over bounded ranges, optimizations can generally assume that the loop induction variable does not wrap. The main wrinkle here is that the form of the loop might be written in a way that requires a *sentinel value*. This comes up if the loop follows typical C style, where the loop condition is checked after an increment.

The Chapel spec currently requires that *sentinel values* are representable:

> An attempt to iterate over a range causes an error if adding stride to the range's last index overflows its index type, i.e. if the sum is greater than the index type's maximum value, or smaller than its minimum value.

While it would be possible to write the loop implementing bounded range iteration without assuming that the *sentinel value* is representable, doing so might add complexity to almost all loops and lead to lower performance in general.

The following loop is an example of bounded range iteration that currently causes a runtime error (without --fast) because the *sentinel value* is not representable:

``` chapel
var a:int(8) = 125;
var b:int(8) = 127;
for i in a..b {
  writeln(i);
}
```

This will currently turn into C like this:
``` c
int8_t a = 125;
int8_t b = 127;
for( int8_t i = a; i <= b; i++ ) {
   writeln(i);
}
```

... and one can imagine that if i++ wraps, i will become `0` and the loop will continue (even though it should not).

A more idomatic C programming way to write a loop is to use `<`. Such a loop structure also needs the sentinel value to be representable, but that fact might be more obvious in this case:
``` c
int i;
for( int i = 125; i < 128; i++ ) {
   writeln(i);
}
```

Nonetheless it might be possible to generate loops like this:
``` c
int8_t a = 125;
int8_t b = 127;
for( int8_t i = a; i <= b; i++ ) {
   writeln(i);
   if ( i == b ) break;
}
```

### Unbounded Range Iteration

For unbounded range iteration, the possibility of overflow is more fundamental and not simply connected to the way in which the loop is represented. For example, consider:

``` chapel
var a:int(8) = 126;
for i in (a..) {
  writeln(i);
}
```

What should this do? Besides the above options:
 * *undefined behavior*
  * *wrap*
  * *trap*
  * *throw*
a new option is possible:
  * *break*: the loop stops iterating before the induction variable would overflow

### Questions to answer

1. What happens generally on signed integer overflow? What about unsigned integer overflow?
2. For bounded range iteration, is overflow an issue? Do we need to continue to assert that a *sentinel value* is representable?
3. What happens with overflow in unbounded range iteration?
4. How can optimizations continue to identify loop induction variables? Or is there a different way to e.g. vectorize

**Comment by mppf** [comment]
Some specific proposals:

"wrap in general, trap/undefined behavior for range iteration"
 1. integer overflow *wraps* generally
 2. same C now; optimization can assume signed induction variables do not wrap
 3. *sentinel value* is still required, 
 4. overflow during range iteration *traps* when bounds checking is on, and is *undefined behavior* when bounds checking is off.

"generate two loops for both safety and performance"
 * integer overflow ?
 * bounded range iteration causes two loops to be generated; bounds checking is performed to choose between the current loop if overflow is not possible and a more general one if it is.


Clarification: in Rust's release mode and our --fast mode, undesirable/buggy overflow *can* happen, whether Rust/we wrap or do something else.

For us, upon --fast we can access invalid memory due to out-of-bounds array indexing. That's in return for better performance. Allowing undesirable overflow upon --fast follows the same philosophy. Therefore this is what I propose that we do.

Aside: when iterating over an unbounded range, we actually iterate until the max value of the range type. I think. At least without --fast.

**Comment by bradcray** [comment]
As with many GitHub issues, I feel that this one tackles several questions which, though thematically related, might be easier to handle as distinct issues.  

Question 3 strikes me as low-hanging fruit that might be quickly resolved as its own issue.  Since we know at compile-time that a range is unbounded, it seems to me that we should be able to generate a loop form that avoids relying on overflowing the high bound by surprise (?).  I'd advocate for having it throw or trap if it does, on the argument that a loop over an unbounded range that runs into the limit of the idxType's representation is a programmer error.  And my guess is that there isn't a performance challenge to this (?).

For normal integer overflow (question 1), it's hard for me to imagine that any of the solutions other than the status quo would avoid introducing a performance penalty that would be regrettable... is that wrong?  For some time, we've discussed a "--check-value-overflows" type of flag that would introduce value checks for cases like this (and downcasts as a second example) as a means of helping the programmer find such bugs in their program while opting into the resulting runtime overhead.

For bounded range iteration, I haven't had any regrets about our status quo decision as of yet, as fiddly as it is...  We'd talked at some point about permitting the user to request through the range type signature the ability to include maximal values -- where the compiler would generate special loops to handle such cases and I still think that's an intriguing idea, though I don't think we ever pursued it.

For #4, the status quo seems attractive to me.  That said, I don't think I understand the distinction implied by the different wordings in the other two solutions.

**Comment by dmk42** [comment]
I'm satisfied that the issue has received attention and the options that have the most merit have been discussed.  Yes, there is a performance penalty associated with not continuing with the status quo.  The worst case is on the order of 20% for full checking of all integer overflows, though the other cases would have some impact too.

Another advantage of choosing the status quo is that if we do, then #6046 can proceed and LLVM vectorization can improve.

As long as it's a conscious choice, I'm fine with keeping the status quo.

**Comment by mppf** [comment]
Responding to some of Brad's points:
> ... might be easier to handle as distinct issues

It might just be me, but I'm uncomfortable answering these questions in isolation. I view the relationships between the answers as an important part of any coherent solution.

> For normal integer overflow (question 1), it's hard for me to imagine that any of the solutions other than the status quo would avoid introducing a performance penalty that would be regrettable... is that wrong?

The actual machine instructions all wrap. So wrapping doesn't have a performance overhead other than the impact upon optimization. If the key cases of optimization are iteration over ranges (as expressed in C for loops in our AST), we can simply make sure that within the context of range iteration / C for loop, our language can provide the optimizations the property they want (that they can assume no overflow can occur in the induction variable).
(But that assumes that what C compilers gain by signed-overflow-is-UB is all about identifying induction variables. That's the case we've identified, but I can't say if there are others that we havn't brought up).

> For #4, the status quo seems attractive to me. That said, I don't think I understand the distinction implied by the different wordings in the other two solutions.

I'll try to clarify, and I've also fixed the above comment.

"status quo" can assume that in
``` chapel
var i:int;
while i < n {
  i += 1;
}
```
the induction variable `i` cannot overflow.

The other two can't make that assumption about the while loop above, but can for a bounded or un-bounded range iteration (I updated the wording to be the same). (Note that a solution that *wraps* for overflow in unbounded range iteration would be able to assume the induction variable does not overflow only for bounded range iteration. But that case hasn't been proposed by anybody).

**Comment by bradcray** [comment]
The machine instructions are guaranteed to wrap?  I was thinking it was behavioral differences across architectures that caused C to make no guarantees that signed ints would wrap...  That said, as long as we're generating to C, we'll have a hard time making them wrap efficiently, won't we?

I think it's obvious/important that in Chapel ranges shouldn't wrap -- ranges should be monotonic sequences of integers.

Thanks for clarifying range iteration vs. while loops over ints.  I don't have as much of an opinion about the while loop case at present -- it seems less common in Chapel and less likely to be used in crucial inner loop code to me than range iteration.

**Comment by dmk42** [comment]
There are two separate issues.  First, not all machines use two's complement, so "wrapping" means something different on different machines, when applied to signed integers.  Most machines still in use today do use two's complement, though it isn't guaranteed.

Second, it's true that it is difficult to be sure a C compiler back end on a two's complement machine will generate wrapping instructions for signed integers; it might just transform the code into something that behaves differently.  That's a relatively new development, in the last 10-15 years.  It wasn't the original reason for undefined overflow on signed ints.

**Comment by dmk42** [comment]
Oh, one more historical note.  In addition to "wrapping" meaning something different depending on the signed encoding, the C standard also wanted to be compatible with the Language-Independent Arithmetic standard, which has a version of integers that trap.  Some hardware actually used to implement traps on signed integer overflow, though I don't know of any still in use today.  (One example, the VAX had a mode that trapped on signed integer overflow, and some people used it.)

**Comment by mppf** [comment]
Re wrapping and 2's complement... Wrapping is the "default", for LLVM IR (which is what started this conversation). You have to explicitly say e.g. `add` won't wrap/overflow if that is what the language guarantees. By default it has this behavior:

> If the sum has unsigned overflow, the result returned is the mathematical result modulo 2n, where n is the bit width of the result.
> Because LLVM integers use a two’s complement representation, this instruction is appropriate for both signed and unsigned integers.

In order words LLVM IR locks in the 2's complement behavior.

I really don't think anybody cares about hardware that doesn't use two's complement. Especially not if we are using the LLVM backend.

I suspect that with casting between signed and unsigned integers we could get C to do this too with little or no overhead. The wrapping behavior in C is well defined for unsigned integers (as I understand it, anyway).

**Comment by dmk42** [comment]
Yes, wrapping is well defined on unsigned integers.  And going straight to LLVM does provide extra flexibility that we can't get when targeting C.  We know exactly what LLVM will do because we control the real input.

**Comment by mppf** [comment]
As far as I understand it at the moment, we're leaning towards "status quo".

**Comment by bradcray** [comment]
If I had a rainy day, I'd lean pretty heavily toward making iteration over unbounded ranges halt/throw on overflow.  It seems it would be unlikely to add overhead to me (given that they'd have to contain something like a conditional + break/return statement to be legal anyway) and seems worthwhile.

**Comment by mppf** [comment]
It sounds to me like the next step here is to update the language specification with "status quo" and further consider the option of unbounded ranges raising some sort of error on overflow (halt, throw, trap...).

**Comment by mppf** [comment]
We are seeing failures with `library/packages/Sort/correctness/sortAllIdxTypes` with `--fast` after enabling the LLVM backend by default.

Here is a small reproducer

``` chapel
proc printx() {
  var rr = -63:int(8)..63:int(8);
  var A:[-63..63] int;
  for i in rr {
    A[i] = rr.size - i;
  }
  writeln(A);
}

printx();
```

This program has different behavior with `--llvm` vs with `--llvm --fast`.

Given the above discussion seemed to be leaning towards "undefined behavior" for signed integer overflow (which this example relies upon), I am planning to modify the test sortAllIdxTypes to avoid signed integer overflow.

**Comment by bradcray** [comment]
@mppf: With only a quick look:  If `.size` returned an `int`, as in #17628 (which also makes some updates to that test), would any change be required here?  If not, should we just wait for that PR to be merged?  (it's just waiting on a review and for LLVM testing to quiet down).

**Comment by bradcray** [comment]
Oops, I see I'm too late.

**Comment by mppf** [comment]
I think we are settled on "status quo" here:

> "status quo"
> 
>     1. unsigned integer overflow _wraps_, signed integer overflow is _undefined behavior_
> 
>     2. _sentinel value_ is required for bounded range iteration; overflow there _traps_ when bounds checking is on, and is _undefined behavior_ when bounds checking is off.
> 
>     3. overflow in unbounded range iteration is _undefined behavior_
> 
>     4. as with C now, optimizations can assume signed induction variables do not wrap.

There was a suggestion that Option 3 might be a separate issue and issue #16419 discusses what should happen with unbounded range iteration.

The remaining TODO here is to update the language specification.
